### ISSUE TITLE
Default font weight chosen by fontselect for PingFang is way too thin

### ISSUE THREAD

**Issue by bodayw** [libass/libass#207]
This has been the case after Core Text is implemented.

PingFang is now the default system Chinese font since OS X 10.11, which is also the fallback font for Chinese characters (if Chinese is set on a higher priority than Japanese of course). While usually the default font weight (e.g. ``Regular`` or ``Normal``) should be chosen, it looks like libass always choose one of the lighter font weight for PingFang (could be ``Light`` or ``Thin``, I can't tell by eye).

Latin fonts look normal here. I also tested some other common (and old) Chinese fonts, like HeiTi and Hiragino Sans GB and they all look normal.

I'm using the mpv builds from @ChrisK2 and I'm not sure if the builds already include the latest changes have been made here.

**Comment by grigorig** [comment]
fontselect internally uses the weight of the font for matching, it's just not exposed to the outside world. Try to tune the weight mapping in https://github.com/libass/libass/blob/master/libass/ass_coretext.c#L119

**Comment by bodayw** [comment]
Hmm...Maybe the fallback is irrelevant here. I just tried adding ``--ass-force-style=FontName=PingFang SC`` in my mpv.conf file and got very thin font for all the subtitle lines. Again, other fonts look normal. Maybe somebody with a Mac can test this? Just use any ASS sub file and force libass to use one of the PingFang font on OSX 10.11.

It's weird that this problem only happens with this particular font family.

**Comment by grigorig** [comment]
Yes, matching doesn't imply fallback. libass needs to sort the faces of a font family according to weight and pick out a suitable one. Something appears to go wrong here.

**Comment by bodayw** [comment]
Any update on this?

**Comment by grigorig** [comment]
Someone with OS X needs to take a look at this. So, not me, unfortunately.

**Comment by astiob** [comment]
These IRC quotes are relevant, I think:

> --- Day changed Thu Dec 31 2015
04:53:11 < Plorkyeran> so uh, the thing the coretext font selector tries to do of converting the weight from coretext's scale to ASS scale doesn't actually work
04:53:29 < Plorkyeran> I made a chart of how the two weights correlate: http://i.imgur.com/9fXAkUe.png
04:53:39 < Plorkyeran> you'll note that they do not
04:54:07 < rcombs> (shouldn't that be a scatter plot)
04:54:39 < Plorkyeran> if there was actually a strict mapping it would be a monotonically increasing line
04:55:07 < Plorkyeran> but sure, a scatter plot is better
04:55:40 < Plorkyeran> http://i.imgur.com/nQi3ktI.png
04:55:59 < rcombs> quality
04:56:53 < Plorkyeran> also coretext sometimes fails to report that a font is italic, but that only actually affects 3/424 font families I have lying around
04:57:10 < Plorkyeran> while the weights being fucked affects nearly all of them

> --- Day changed Mon Jan 04 2016
02:42:51 < Plorkyeran> I wrote a gross mess that matches GDI's font selection on nearly everything I have to throw at it (all but 40 of 7644 family+weight+italic combinations)
02:42:54 < Plorkyeran> https://github.com/Aegisub/Aegisub/blob/master/src/font_file_lister_coretext.mm
02:43:18 < Plorkyeran> I plan to take a stab at improving libass's matching sometime, but I'm kinda sick of dealing with fonts for now

(The time zone is Europe/London.)

(@tgoyne @rcombs I hope you don’t mind this being posted. Let me know if you do.)

**Comment by astiob** [comment]
Basically, as far as I can tell, we need to stop believing most of the metadata Core Text gives us and instead ask it for the raw tables and decode them ourselves.

**Comment by tgoyne** [comment]
From memory, the things that libass's coretext provider does wrong (where "wrong" is defined as "different from GDI") are:

1. GDI reduces the weight penalty for fonts that will be emboldened due to being sufficiently too light. IIRC this ended up not actually mattering for any real fonts.
2. GDI and CoreText read different fields for font weights, and font designers are not very good at making them match.
3. GDI and CoreText read different fields for font slants, but unlike with weights they almost always do match, so using CoreText's value is fine.
4. GDI and CoreText read different fields for all of the names (if both are present). Unlike with weight the only places I observed differences here due to incompetence were with shitty fansubber-modified fonts, but there's an unfortunately common pattern where the non-regular/bold/itatlic/bolditatlic styles of a font include the style name in the microsoft family name (since windows 3.1 didn't directly support any other variants), but not in the macintosh family name, since Mac OS always did support arbitrary variants. This actually mostly works out okay, except for that CoreText doesn't always pick the correct variant when given a bare family name.
5. CoreText refused to match a few font files that I have lying around that it could correctly render text with when told to use that file directly, and which font book would display correctly. I'm not sure why.

IIRC I concluded that despite the problems CoreText's matching was useful in that matching on the ASS-provided font name gave a superset of the correct fonts for all but the most broken cases, and filtering that set to the correct set was sufficiently fast to not need a separate index, while enumerating all installed fonts and doing the filtering entirely myself would definitely require an index and all the fun that comes with that.

I went with manually parsing the data structures over using freetype because I couldn't figure out how to handle TTCs correctly with freetype (IIRC figuring out which font in the collection coretext had matched was way more code than the half-assed parsing).

**Comment by grigorig** [comment]
Handling TTCs is pretty simple - you can provide a a font face index when loading a font face from a TTC. There should be some simple code in libass already to deal with CoreText's way which isn't based on these indices. Or maybe I don't really understand the question.

**Comment by tgoyne** [comment]
The font descriptor supplied by CoreText doesn't include the font face index (unless it's another one of those things they just didn't bother mentioning in the documentation).

**Comment by grigorig** [comment]
Yes, that is why libass uses the PostScript name to find the correct index.

**Comment by bodayw** [comment]
Well...Just realized that the bug I described above is already fixed, at least that is what it seems like.

No idea about when and or change, which was either made here or on mpv's side, fixed the issue. It could be the recent release 0.13.2, though I'm not sure...

Feel free to close this. I'm happy with the result right now.

**Comment by grigorig** [comment]
That is odd, there haven't really been any changes to the CoreText backend. Maybe the font changed with an updated OSX version?

**Comment by bodayw** [comment]
>Maybe the font changed with an updated OSX version?

It's possible, but I didn't track the font file changes so there is no way for me to know.
According to [http://www.jklstudios.com/misc/osxfonts.html](url) PingFang and quite a few of other fonts were updated in 10.11.2, which was already released last December, so it can't be those. I haven't found other information regarding recent OS X updates.

I'm no expert on this topic, but if I understand correctly, that issue raised by @astiob earlier here is more of a general one which could affect any font, while my original problem only happens with several of the new fonts introduced in OS X 10.11. If Apple did change those fonts in a recent system update, I guess they just somehow improved the compatibility so that CoreText, thus libass could work much better with them, although it could still well be that the font weight are not selected in a "precise" or optimal way, as suggested above by @tgoyne.
